packages/orchestrator/lib/routes/v1/dequeue.ts

```diff
@@ -0,0 +1,90 @@
+import { z } from 'zod';
+import type { Scheduler, Task } from '@nangohq/scheduler';
+import type { ApiError, Endpoint } from '@nangohq/types';
+import type { EndpointRequest, EndpointResponse, RouteHandler, Route } from '@nangohq/utils';
+import { validateRequest } from '@nangohq/utils';
+import type EventEmitter from 'node:events';
+
+const path = '/v1/dequeue';
+const method = 'GET';
+
+type GetDequeue = Endpoint<{
+    Method: typeof method;
+    Path: typeof path;
+    Querystring: {
+        groupKey: string;
+        limit: number;
+        waitForCompletion?: boolean;
+    };
+    Error: ApiError<'dequeue_failed'>;
+    Success: Task[];
+}>;
+
+const validate = validateRequest<GetDequeue>({
+    parseQuery: (data) =>
+        z
+            .object({
+                groupKey: z.string().min(1),
+                limit: z.coerce.number().positive(),
+                waitForCompletion: z
+                    .string()
+                    .optional()
+                    .default('false')
+                    .transform((val) => val === 'true')
+            })
+            .parse(data)
+});
+
+export const getRoute: Route<GetDequeue> = { path, method };
+
+export const getRouteHandler = (scheduler: Scheduler, eventEmitter: EventEmitter): RouteHandler<GetDequeue> => {
+    return {
+        ...getRoute,
+        validate,
+        handler: getHandler(scheduler, eventEmitter)
+    };
+};
+
+const getHandler = (scheduler: Scheduler, eventEmitter: EventEmitter) => {
+    return async (req: EndpointRequest<GetDequeue>, res: EndpointResponse<GetDequeue>) => {
+        const { groupKey, limit } = req.query;
+        const waitForCompletionTimeoutMs = 60_000;
+        const eventId = `task:started:${groupKey}`;
+        const cleanupAndRespond = (respond: (res: EndpointResponse<GetDequeue>) => void) => {
+            if (timeout) {
+                clearTimeout(timeout);
+            }
+            if (onTaskStarted) {
+                eventEmitter.removeListener(eventId, onTaskStarted);
+            }
+            if (!res.writableEnded) {
+                respond(res);
+            }
+        };
+        const onTaskStarted = async (_t: Task) => {
+            const getTasks = await scheduler.dequeue({ groupKey, limit });
+            if (getTasks.isErr()) {
+                cleanupAndRespond((res) => res.status(500).json({ error: { code: 'dequeue_failed', message: getTasks.error.message } }));
+            } else {
+                cleanupAndRespond((res) => res.status(200).json(getTasks.value));
+            }
+        };
+        const timeout = setTimeout(() => {
+            cleanupAndRespond((res) => res.status(200).send([]));
+        }, waitForCompletionTimeoutMs);
+
+        eventEmitter.once(eventId, onTaskStarted);
```

Comment:
not sure I get the wait for completion in that case, because if don't receive an answer you'll just wait 60s for nothing

Reply:
you are waiting for a task to be created/dequeuable, so you don't have to keep making request.

Reply:
okay took me a while to understand the executeAction -> execute -> schedule -> immediate -> event 
that works 👍🏻